File: NEsper.Tests/eql/spec/TestPatternStreamSpecRaw.cs
// ---------------------------------------------------------------------------------- /
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
// ---------------------------------------------------------------------------------- /

using System;
using System.Collections.Generic;

using NUnit.Framework;

using antlr.collections;

using net.esper.eql.core;
using net.esper.eql.expression;
using net.esper.eql.parse;
using net.esper.filter;
using net.esper.pattern;
using net.esper.support.bean;
using net.esper.support.eql.parse;
using net.esper.support.events;

namespace net.esper.eql.spec
{
	[TestFixture]
	public class TestPatternStreamSpecRaw
	{
	    [Test]
	    public void testPatternEquals()
	    {
	        String text = "select * from pattern [" +
	                "s=" + typeof(SupportBean).FullName + "(intPrimitive=5) -> " +
	                "t=" + typeof(SupportBean).FullName + "(intPrimitive=s.intBoxed)" +
	                "]";
	        TryPatternEquals(text);

	        text = "select * from pattern [" +
	                "s=" + typeof(SupportBean).FullName + "(5=intPrimitive) -> " +
	                "t=" + typeof(SupportBean).FullName + "(s.intBoxed=intPrimitive)" +
	                "]";
	        TryPatternEquals(text);
	    }

	    [Test]
	    public void testInvalid()
	    {
	        String text = "select * from pattern [" +
	                "s=" + typeof(SupportBean).FullName + " -> " +
	                "t=" + typeof(SupportBean).FullName + "(intPrimitive=s.doubleBoxed)" +
	                "]";
	        TryInvalid(text);

	        text = "select * from pattern [" +
	                "s=" + typeof(SupportBean).FullName + " -> " +
	                "t=" + typeof(SupportBean).FullName + "(intPrimitive in (s.doubleBoxed))" +
	                "]";
	        TryInvalid(text);
	    }

	    private void TryInvalid(String text)
	    {
	        try
	        {
	            PatternStreamSpecRaw raw = MakeSpec(text);
	            Compile(raw);
	            Assert.Fail();
	        }
	        catch (ExprValidationException ex)
	        {
	            // expected
	        }
	    }

	    [Test]
	    public void testPatternExpressions()
	    {
	        String text = "select * from pattern [" +
	                "s=" + typeof(SupportBean).FullName + "(intPrimitive in (s.intBoxed + 1, 0), intBoxed+1=intPrimitive-1)" +
	                "]";

	        PatternStreamSpecRaw raw = MakeSpec(text);
	        PatternStreamSpecCompiled spec = Compile(raw);
	        Assert.AreEqual(1, spec.TaggedEventTypes.Count);
	        Assert.AreEqual(typeof(SupportBean), spec.TaggedEventTypes.Fetch("s").UnderlyingType);

	        List<EvalFilterNode> filters = EvalNode.RecusiveFilterChildNodes(spec.EvalNode);
	        Assert.AreEqual(1, filters.Count);

	        // node 0
	        EvalFilterNode filterNode = filters[0];
	        Assert.AreEqual(typeof(SupportBean), filterNode.FilterSpec.EventType.UnderlyingType);
	        Assert.AreEqual(1, filterNode.FilterSpec.Parameters.Count);
	        FilterSpecParamExprNode exprParam = (FilterSpecParamExprNode) filterNode.FilterSpec.Parameters[0];
	    }

	    [Test]
	    public void testPatternInSetOfVal()
	    {
	        String text = "select * from pattern [" +
	                "s=" + typeof(SupportBean).FullName + " -> " +
	                       typeof(SupportBean).FullName + "(intPrimitive in (s.intBoxed, 0))" +
	                "]";

	        PatternStreamSpecRaw raw = MakeSpec(text);
	        PatternStreamSpecCompiled spec = Compile(raw);
	        Assert.AreEqual(1, spec.TaggedEventTypes.Count);
	        Assert.AreEqual(typeof(SupportBean), spec.TaggedEventTypes.Fetch("s").UnderlyingType);

	        List<EvalFilterNode> filters = EvalNode.RecusiveFilterChildNodes(spec.EvalNode);
	        Assert.AreEqual(2, filters.Count);

	        // node 0
	        EvalFilterNode filterNode = filters[0];
	        Assert.AreEqual(typeof(SupportBean), filterNode.FilterSpec.EventType.UnderlyingType);
	        Assert.AreEqual(0, filterNode.FilterSpec.Parameters.Count);

	        // node 1
	        filterNode = filters[1];
	        Assert.AreEqual(typeof(SupportBean), filterNode.FilterSpec.EventType.UnderlyingType);
	        Assert.AreEqual(1, filterNode.FilterSpec.Parameters.Count);

	        FilterSpecParamIn inlist = (FilterSpecParamIn) filterNode.FilterSpec.Parameters[0];
	        Assert.AreEqual(FilterOperator.IN_LIST_OF_VALUES, inlist.FilterOperator);
	        Assert.AreEqual(2, inlist.ListOfValues.Count);

	        // in-value 1
	        InSetOfValuesEventProp prop = (InSetOfValuesEventProp) inlist.ListOfValues[0];
	        Assert.AreEqual("s", prop.ResultEventAsName);
	        Assert.AreEqual("intBoxed", prop.ResultEventProperty);

	        // in-value 1
	        InSetOfValuesConstant constant = (InSetOfValuesConstant) inlist.ListOfValues[1];
	        Assert.AreEqual(0, constant.GetConstant());
	    }

	    [Test]
	    public void testRange()
	    {
	        String text = "select * from pattern [" +
	                "s=" + typeof(SupportBean).FullName + " -> " +
	                       typeof(SupportBean).FullName + "(intPrimitive between s.intBoxed and 100)" +
	                "]";

	        PatternStreamSpecRaw raw = MakeSpec(text);
	        PatternStreamSpecCompiled spec = Compile(raw);
	        Assert.AreEqual(1, spec.TaggedEventTypes.Count);
	        Assert.AreEqual(typeof(SupportBean), spec.TaggedEventTypes.Fetch("s").UnderlyingType);

	        List<EvalFilterNode> filters = EvalNode.RecusiveFilterChildNodes(spec.EvalNode);
	        Assert.AreEqual(2, filters.Count);

	        // node 0
	        EvalFilterNode filterNode = filters[0];
	        Assert.AreEqual(typeof(SupportBean), filterNode.FilterSpec.EventType.UnderlyingType);
	        Assert.AreEqual(0, filterNode.FilterSpec.Parameters.Count);

	        // node 1
	        filterNode = filters[1];
	        Assert.AreEqual(typeof(SupportBean), filterNode.FilterSpec.EventType.UnderlyingType);
	        Assert.AreEqual(1, filterNode.FilterSpec.Parameters.Count);

	        FilterSpecParamRange range = (FilterSpecParamRange) filterNode.FilterSpec.Parameters[0];
	        Assert.AreEqual(FilterOperator.RANGE_CLOSED, range.FilterOperator);

	        // min-value
	        RangeValueEventProp prop = (RangeValueEventProp) range.Min;
	        Assert.AreEqual("s", prop.ResultEventAsName);
	        Assert.AreEqual("intBoxed", prop.ResultEventProperty);

	        // max-value
	        RangeValueDouble constant = (RangeValueDouble) range.Max;
	        Assert.AreEqual(100d, constant.DoubleValue);
	    }

	    private void TryPatternEquals(String text)
	    {
	        PatternStreamSpecRaw raw = MakeSpec(text);
	        PatternStreamSpecCompiled spec = Compile(raw);
	        Assert.AreEqual(2, spec.TaggedEventTypes.Count);
	        Assert.AreEqual(typeof(SupportBean), spec.TaggedEventTypes.Fetch("s").UnderlyingType);
	        Assert.AreEqual(typeof(SupportBean), spec.TaggedEventTypes.Fetch("t").UnderlyingType);

	        List<EvalFilterNode> filters = EvalNode.RecusiveFilterChildNodes(spec.EvalNode);
	        Assert.AreEqual(2, filters.Count);

	        // node 0
	        EvalFilterNode filterNode = filters[0];
	        Assert.AreEqual(typeof(SupportBean), filterNode.FilterSpec.EventType.UnderlyingType);
	        Assert.AreEqual(1, filterNode.FilterSpec.Parameters.Count);

	        FilterSpecParamConstant constant = (FilterSpecParamConstant) filterNode.FilterSpec.Parameters[0];
	        Assert.AreEqual(FilterOperator.EQUAL, constant.FilterOperator);
	        Assert.AreEqual("intPrimitive", constant.PropertyName);
	        Assert.AreEqual(5, constant.FilterConstant);

	        // node 1
	        filterNode = filters[1];
	        Assert.AreEqual(typeof(SupportBean), filterNode.FilterSpec.EventType.UnderlyingType);
	        Assert.AreEqual(1, filterNode.FilterSpec.Parameters.Count);

	        FilterSpecParamEventProp eventprop = (FilterSpecParamEventProp) filterNode.FilterSpec.Parameters[0];
	        Assert.AreEqual(FilterOperator.EQUAL, constant.FilterOperator);
	        Assert.AreEqual("intPrimitive", constant.PropertyName);
	        Assert.AreEqual("s", eventprop.ResultEventAsName);
	        Assert.AreEqual("intBoxed", eventprop.ResultEventProperty);
	    }

	    private PatternStreamSpecCompiled Compile(PatternStreamSpecRaw raw)
	    {
	        PatternStreamSpecCompiled compiled = (PatternStreamSpecCompiled) raw.Compile(SupportEventAdapterService.GetService(), new MethodResolutionServiceImpl(new EngineImportServiceImpl()));
	        return compiled;
	    }

	    private static PatternStreamSpecRaw MakeSpec(String expression)
	    {
	        AST ast = SupportParserHelper.ParseEQL(expression);
	        SupportParserHelper.DisplayAST(ast);

	        EQLTreeWalker walker = SupportEQLTreeWalkerFactory.MakeWalker();
            walker.startEQLExpressionRule(ast);

	        PatternStreamSpecRaw spec = (PatternStreamSpecRaw) walker.StatementSpec.StreamSpecs[0];
	        return spec;
	    }
	}
} // End of namespace

File: trunk/esper/src/test/java/net/esper/eql/spec/TestPatternStreamSpecRaw.java
package net.esper.eql.spec;

import antlr.collections.AST;
import junit.framework.TestCase;
import net.esper.eql.core.MethodResolutionServiceImpl;
import net.esper.eql.core.EngineImportServiceImpl;
import net.esper.eql.parse.EQLTreeWalker;
import net.esper.eql.expression.ExprValidationException;
import net.esper.filter.*;
import net.esper.pattern.EvalFilterNode;
import net.esper.pattern.EvalNode;
import net.esper.pattern.EvalNodeAnalysisResult;
import net.esper.support.bean.SupportBean;
import net.esper.support.eql.parse.SupportParserHelper;
import net.esper.support.eql.parse.SupportEQLTreeWalkerFactory;
import net.esper.support.event.SupportEventAdapterService;

import java.util.List;

public class TestPatternStreamSpecRaw extends TestCase
{
    public void testPatternEquals() throws Exception
    {
        String text = "select * from pattern [" +
                "s=" + SupportBean.class.getName() + "(intPrimitive=5) -> " +
                "t=" + SupportBean.class.getName() + "(intPrimitive=s.intBoxed)" +
                "]";
        tryPatternEquals(text);

        text = "select * from pattern [" +
                "s=" + SupportBean.class.getName() + "(5=intPrimitive) -> " +
                "t=" + SupportBean.class.getName() + "(s.intBoxed=intPrimitive)" +
                "]";
        tryPatternEquals(text);
    }

    public void testInvalid() throws Exception
    {
        String text = "select * from pattern [" +
                "s=" + SupportBean.class.getName() + " -> " +
                "t=" + SupportBean.class.getName() + "(intPrimitive=s.doubleBoxed)" +
                "]";
        tryInvalid(text);

        text = "select * from pattern [" +
                "s=" + SupportBean.class.getName() + " -> " +
                "t=" + SupportBean.class.getName() + "(intPrimitive in (s.doubleBoxed))" +
                "]";
        tryInvalid(text);
    }

    private void tryInvalid(String text) throws Exception
    {
        try
        {
            PatternStreamSpecRaw raw = makeSpec(text);
            compile(raw);
            fail();
        }
        catch (ExprValidationException ex)
        {
            // expected
        }
    }

    public void testPatternExpressions() throws Exception
    {
        String text = "select * from pattern [" +
                "s=" + SupportBean.class.getName() + "(intPrimitive in (s.intBoxed + 1, 0), intBoxed+1=intPrimitive-1)" +
                "]";

        PatternStreamSpecRaw raw = makeSpec(text);
        PatternStreamSpecCompiled spec = compile(raw);
        assertEquals(1, spec.getTaggedEventTypes().size());
        assertEquals(SupportBean.class, spec.getTaggedEventTypes().get("s").getUnderlyingType());

        EvalNodeAnalysisResult evalNodeAnalysisResult = EvalNode.recursiveAnalyzeChildNodes(spec.getEvalNode());
        List<EvalFilterNode> filters = evalNodeAnalysisResult.getFilterNodes();
        assertEquals(1, filters.size());

        // node 0
        EvalFilterNode filterNode = filters.get(0);
        assertEquals(SupportBean.class, filterNode.getFilterSpec().getEventType().getUnderlyingType());
        assertEquals(1, filterNode.getFilterSpec().getParameters().size());
        FilterSpecParamExprNode exprParam = (FilterSpecParamExprNode) filterNode.getFilterSpec().getParameters().get(0);   
    }

    public void testPatternInSetOfVal() throws Exception
    {
        String text = "select * from pattern [" +
                "s=" + SupportBean.class.getName() + " -> " +
                       SupportBean.class.getName() + "(intPrimitive in (s.intBoxed, 0))" +
                "]";

        PatternStreamSpecRaw raw = makeSpec(text);
        PatternStreamSpecCompiled spec = compile(raw);
        assertEquals(1, spec.getTaggedEventTypes().size());
        assertEquals(SupportBean.class, spec.getTaggedEventTypes().get("s").getUnderlyingType());

        EvalNodeAnalysisResult evalNodeAnalysisResult = EvalNode.recursiveAnalyzeChildNodes(spec.getEvalNode());
        List<EvalFilterNode> filters = evalNodeAnalysisResult.getFilterNodes();
        assertEquals(2, filters.size());

        // node 0
        EvalFilterNode filterNode = filters.get(0);
        assertEquals(SupportBean.class, filterNode.getFilterSpec().getEventType().getUnderlyingType());
        assertEquals(0, filterNode.getFilterSpec().getParameters().size());

        // node 1
        filterNode = filters.get(1);
        assertEquals(SupportBean.class, filterNode.getFilterSpec().getEventType().getUnderlyingType());
        assertEquals(1, filterNode.getFilterSpec().getParameters().size());

        FilterSpecParamIn inlist = (FilterSpecParamIn) filterNode.getFilterSpec().getParameters().get(0);
        assertEquals(FilterOperator.IN_LIST_OF_VALUES, inlist.getFilterOperator());
        assertEquals(2, inlist.getListOfValues().size());

        // in-value 1
        InSetOfValuesEventProp prop = (InSetOfValuesEventProp) inlist.getListOfValues().get(0);
        assertEquals("s", prop.getResultEventAsName());
        assertEquals("intBoxed", prop.getResultEventProperty());

        // in-value 1
        InSetOfValuesConstant constant = (InSetOfValuesConstant) inlist.getListOfValues().get(1);
        assertEquals(0, constant.getConstant());
    }

    public void testRange() throws Exception
    {
        String text = "select * from pattern [" +
                "s=" + SupportBean.class.getName() + " -> " +
                       SupportBean.class.getName() + "(intPrimitive between s.intBoxed and 100)" +
                "]";

        PatternStreamSpecRaw raw = makeSpec(text);
        PatternStreamSpecCompiled spec = compile(raw);
        assertEquals(1, spec.getTaggedEventTypes().size());
        assertEquals(SupportBean.class, spec.getTaggedEventTypes().get("s").getUnderlyingType());

        EvalNodeAnalysisResult evalNodeAnalysisResult = EvalNode.recursiveAnalyzeChildNodes(spec.getEvalNode());
        List<EvalFilterNode> filters = evalNodeAnalysisResult.getFilterNodes();
        assertEquals(2, filters.size());

        // node 0
        EvalFilterNode filterNode = filters.get(0);
        assertEquals(SupportBean.class, filterNode.getFilterSpec().getEventType().getUnderlyingType());
        assertEquals(0, filterNode.getFilterSpec().getParameters().size());

        // node 1
        filterNode = filters.get(1);
        assertEquals(SupportBean.class, filterNode.getFilterSpec().getEventType().getUnderlyingType());
        assertEquals(1, filterNode.getFilterSpec().getParameters().size());

        FilterSpecParamRange range = (FilterSpecParamRange) filterNode.getFilterSpec().getParameters().get(0);
        assertEquals(FilterOperator.RANGE_CLOSED, range.getFilterOperator());

        // min-value
        RangeValueEventProp prop = (RangeValueEventProp) range.getMin();
        assertEquals("s", prop.getResultEventAsName());
        assertEquals("intBoxed", prop.getResultEventProperty());

        // max-value
        RangeValueDouble constant = (RangeValueDouble) range.getMax();
        assertEquals(100d, constant.getDoubleValue());
    }

    private void tryPatternEquals(String text) throws Exception
    {
        PatternStreamSpecRaw raw = makeSpec(text);
        PatternStreamSpecCompiled spec = compile(raw);
        assertEquals(2, spec.getTaggedEventTypes().size());
        assertEquals(SupportBean.class, spec.getTaggedEventTypes().get("s").getUnderlyingType());
        assertEquals(SupportBean.class, spec.getTaggedEventTypes().get("t").getUnderlyingType());

        EvalNodeAnalysisResult evalNodeAnalysisResult = EvalNode.recursiveAnalyzeChildNodes(spec.getEvalNode());
        List<EvalFilterNode> filters = evalNodeAnalysisResult.getFilterNodes();
        assertEquals(2, filters.size());

        // node 0
        EvalFilterNode filterNode = filters.get(0);
        assertEquals(SupportBean.class, filterNode.getFilterSpec().getEventType().getUnderlyingType());
        assertEquals(1, filterNode.getFilterSpec().getParameters().size());

        FilterSpecParamConstant constant = (FilterSpecParamConstant) filterNode.getFilterSpec().getParameters().get(0);
        assertEquals(FilterOperator.EQUAL, constant.getFilterOperator());
        assertEquals("intPrimitive", constant.getPropertyName());
        assertEquals(5, constant.getFilterConstant());

        // node 1
        filterNode = filters.get(1);
        assertEquals(SupportBean.class, filterNode.getFilterSpec().getEventType().getUnderlyingType());
        assertEquals(1, filterNode.getFilterSpec().getParameters().size());

        FilterSpecParamEventProp eventprop = (FilterSpecParamEventProp) filterNode.getFilterSpec().getParameters().get(0);
        assertEquals(FilterOperator.EQUAL, constant.getFilterOperator());
        assertEquals("intPrimitive", constant.getPropertyName());
        assertEquals("s", eventprop.getResultEventAsName());
        assertEquals("intBoxed", eventprop.getResultEventProperty());
    }

    private PatternStreamSpecCompiled compile(PatternStreamSpecRaw raw) throws Exception
    {
        PatternStreamSpecCompiled compiled = (PatternStreamSpecCompiled) raw.compile(SupportEventAdapterService.getService(), new MethodResolutionServiceImpl(new EngineImportServiceImpl()), null);
        return compiled;
    }

    private static PatternStreamSpecRaw makeSpec(String expression) throws Exception
    {
        AST ast = SupportParserHelper.parseEQL(expression);
        SupportParserHelper.displayAST(ast);

        EQLTreeWalker walker = SupportEQLTreeWalkerFactory.makeWalker();        
        walker.startEQLExpressionRule(ast);

        PatternStreamSpecRaw spec = (PatternStreamSpecRaw) walker.getStatementSpec().getStreamSpecs().get(0);
        return spec;
    }
}
